library/alloc/src/collections/btree/map.rs

```diff
@@ -167,6 +144,43 @@ pub(super) const MIN_LEN: usize = node::MIN_LEN_AFTER_SPLIT;
 /// // modify an entry before an insert with in-place mutation
 /// player_stats.entry("mana").and_modify(|mana| *mana += 200).or_insert(100);
 /// ```
+///
+/// # Background
+///
+/// A B-tree is (like) a [binary search tree], but adapted to the natural granularity that modern machines like to consume data at.
+/// This means that each node contains an entire array of elements, instead of just a single element.
+///
+/// B-Trees represent a fundamental compromise between cache-efficiency and actually minimizing
+/// the amount of work performed in a search. In theory, a binary search tree (BST) is the optimal
+/// choice for a sorted map, as a perfectly balanced BST performs the theoretical minimum number of
+/// comparisons necessary to find an element (log<sub>2</sub>n). However, in practice the way this
+/// is done is *very* inefficient for modern computer architectures. In particular, every element
+/// is stored in its own individually heap-allocated node. This means that every single insertion
+/// triggers a heap-allocation, and for every comparison a node needs to be loaded,
+/// which could result in a cache miss. Since both heap-allocations and cache-misses
```

Comment:
```suggestion
/// triggers a heap-allocation, and comparison is a potential cache-miss due to the indirection.
/// Since both heap-allocations and cache-misses
```